src/stream/src/executor/dml.rs

```diff
@@ -151,8 +166,35 @@ impl DmlExecutor {
                         TxnMsg::End(txn_id) => {
                             let mut txn_buffer = active_txn_map.remove(&txn_id)
                                 .unwrap_or_else(|| panic!("Receive an unexpected transaction end message. Active transaction map doesn't contain this transaction txn_id = {}.", txn_id));
-                            for chunk in txn_buffer.vec.drain(..) {
-                                yield Message::Chunk(chunk);
+
+                            let txn_buffer_cardinality = txn_buffer
+                                .vec
+                                .iter()
+                                .map(|c| c.cardinality())
+                                .sum::<usize>();
+                            let batch_group_cardinality =
+                                batch_group.iter().map(|c| c.cardinality()).sum::<usize>();
+
+                            if txn_buffer_cardinality >= self.chunk_size {
+                                // txn buffer isn't small, so yield.
+                                for chunk in txn_buffer.vec.drain(..) {
+                                    yield Message::Chunk(chunk);
+                                }
+                            } else if txn_buffer_cardinality + batch_group_cardinality
+                                <= self.chunk_size
+                            {
+                                // txn buffer is small and batch group has space.
+                                for chunk in txn_buffer.vec.drain(..) {
+                                    batch_group.push(chunk);
+                                }
+                            } else {
+                                // txn buffer is small and batch group has no space, so yield the large one.
+                                if txn_buffer_cardinality < batch_group_cardinality {
+                                    mem::swap(&mut txn_buffer.vec, &mut batch_group);
+                                }
+                                let concat_chunk = StreamChunk::concat(txn_buffer.vec);
```

Comment:
If there're remaining spaces in the "txn buffer", should we also yield some chunks from the "batch group" here?

Can we utilize `StreamChunkBuilder` for doing this all stuff?

Reply:
If we use that space I am afraid it would make small transaction data interleaved with each other.

Reply:
That's true. Will this lead to problems?

Reply:
Depends on the downstream table conflicts handling logic. I think the default policy is the first writer win, so it is possible to cause some undetermined behavior under concurrent writes. Prefer to keep this invariant.

Reply:
I see. However if I understand correctly, records written by transactions from different sessions are already interleaved by the shuffle in the downstream. 🤔

Reply:
Yes, the hash shuffle in the downstream is unavoidable, but I don't want to introduce another variable.